src/jit/optcse.cpp

```diff
@@ -2574,11 +2574,29 @@ bool Compiler::optIsCSEcandidate(GenTree* tree)
         return false;
     }
 
-    // If this is a struct type, we can only consider it for CSE-ing if we can get at
-    // its handle, so that we can create a temp.
-    if ((type == TYP_STRUCT) && (gtGetStructHandleIfPresent(tree) == NO_CLASS_HANDLE))
+    // For some struct types, we now allow them to be CSE candidates.
+    //
+    if (type == TYP_STRUCT)
     {
-        return false;
+        // We can only consider it for CSE-ing if we can get at
+        // its handle, so that we can create a temp.
+        //
+        CORINFO_CLASS_HANDLE structHnd = gtGetStructHandleIfPresent(tree);
+        if (structHnd == NO_CLASS_HANDLE)
+        {
+            return false;
+        }
+
+        unsigned          size = info.compCompHnd->getClassSize(structHnd);
+        structPassingKind howToPassStruct;
+        var_types         structBaseType = getArgTypeForStruct(structHnd, &howToPassStruct, false, size);
+
+        // Only allow structs whose storage can be held in a register to be CSE candidates.
```

Comment:
Presumably that's just because it's not very useful to CSE something that doesn't fit in a register. Or is there a correctness issue as well?

I suppose it would make sense to CSE a struct indir if the new temporary variable can then be promoted. But that's probably quite complicated to get right.

Reply:
Yes, There currently is a correctness issue as well when we perform loop hoisting we place a l-value which is a GT_IND of a TYP_STRUCT into the loop pre-header.
This indirection doesn't get handled properly (or removed) and we end up asserting about a missing register during codegen.

Reply:
see bool Compiler::optHoistLoopExprsForTree()
where we currenmtly require:

```
        // Tree must be a suitable CSE candidate for us to be able to hoist it.
        treeIsHoistable &= optIsCSEcandidate(tree);
```

Reply:
Ah, of course, loop hoisting will hoist the LHS of the struct assignment but that does not have a struct handle so there's no way for it to actually work.

Reply:
Still not clear to me if the above is a legality check followed by a profitability check, or if both are legality checks. Can you clarify?

Reply:
Currently is both a profitability check and a legality check.
This method is used in the hoisting logic and when it returns true we can perform an illegal hoist of a struct l-value.
I don't believe that creating CSE of multi-slot structs is ever going generate better code,
CSE are profitable when the def cost is expensive relative to the use costs.
The def costs for non-enregisterable  multi-slot structs is never going to be that much more expensive than the use cost.  If the def is  a heavy complex expression, such as an array expression we will still perform a CSE on the address calculation, leaving us with a simple multi-slot load as the CSE def. Since it isn't enregisterable  the CSE use still have to load from memory (the local stack frame) so we are just moving a copy of the struct from global memory into a stack copy and not saving any memory operations.

Reply:
@briansull - in our discussions of this, it seemed like the fundamental issues are:
1) The struct handling in the JIT relies on the fact that a struct r-value will always be used in a context where its size and/or handle are carried by an l-value (e.g. a block node on the lhs of an assign, or a block store).
2) An unused hoisted rvalue doesn't get removed. This leaves the problematic handle-less size-less struct indirection.

Before taking this fix, I'd like us to understand what diffs this creates, and file an issue to address the above problems.

As we improve our struct codegen, I'd expect us to see more cases where hoisting struct values would be beneficial, as we can't (and probably never will) promote all structs that might be created redundantly (within a loop or otherwise).

Reply:
> An unused hoisted rvalue doesn't get removed. This leaves the problematic handle-less size-less struct indirection.

Not only that it doesn't get removed but it simply cannot be used at all. Because loop hoisting is expecting CSE to assign the hoisted tree to a variable but CSE can't do that because there's struct type information available in the r-value.

To make this work with the current loop hoisiting implementation we'd need to stop discarding struct information from r-values (converting `OBJ` to `TYP_STRUCT` `IND`). That's doable and IMO desirable, discarding struct information is just risky.

Reply:
FWIW I have seen many examples where manual struct CSEs are profitable -- for example the before version of this did not cache `Start` and `End` as local structs. IL semantics for structs tends to lead to lots of struct copies and often many of those copies are identical.

https://github.com/dotnet/coreclr/blob/0d581c79f0b40bd43ad3eb62574bfe78a665fe80/src/System.Private.CoreLib/shared/System/Range.cs#L105-L122

Reply:
Hmm, should perhaps this fix be restricted to loop hoisting?

Reply:
I  have coded up an alternative fix that blocks the hoisting of TYP_STRUCT r-values.